### DISCUSSION TITLE
autoDiscoverFilter exclude certain subdirectories

### DISCUSSION THREAD

**Discussion by stefandoorn** [renovatebot/renovate#16466] [Request Help]
Hi,

I'm having difficulties doing the following:

* Include all repositories inside a group
* Exclude a certain path from it

To include all, I have this to the RENOVATE_EXTRA_FLAGS env var: `--autodiscover=true --autodiscover-filter=group/**`

This works fine.

Now I want to exclude the `playgrounds` subfolder (group/playgrounds/**), but everything I'm trying doesn't seem to work. I do see it takes a minimatch syntax, but also having trouble with that.

Could someone help me how to exclude this certain subgroup?

I do feel more people struggle with this, so I'm happy to suggest a documentation change once I fully understand this.

**Comment by PhilipAbed** [comment]
did you try to use the regex instead of the minimatch?
> All text inside the start and end / will be treated as a regular expression.
```
{
  "autodiscoverFilter": "/your regex/"
}
```
i think with regex you can exclude some of them.

but in general, i think this should be an Array of regexes, so regex wont get too complicated, and people will be able to choose specific groups of folders.

**Reply by stefandoorn** [reply]
I tried a lot, as well some regexpes. 

But, I'll try some more and also write down what didn't work. I just kept trying and trying, without noting down what didn't work. 

An array would be great indeed, I think that makes it more easy for many.

**Reply by stefandoorn** [reply]
I also overlooked that the `regexp` should be enclosed. I read it in the docs as if that enclosing wasn't needed. But, yeah, that wasn't the most handy assumption I think.

**Reply by viceice** [reply]
we needed to revert array style because of breakings on ENV or cli config

**Reply by PhilipAbed** [reply]
@stefandoorn  play a bit with the regex as i just made a POC, 
check this https://regex101.com/r/66OINf/1
you might need to add g at the end i think to make it global `/{regex}/g`  
Update: no need for the `g` flag i believe try without

**Reply by PhilipAbed** [reply]
@viceice the alternative is to create an Exclude filter to the auto discover for complicated needs, not sure its necessary right now but some people might need it in the future , do we have any other way to exclude from the filtered repositories?

**Reply by viceice** [reply]
no global flag set and allowed for regex, i think you should test with single line flag when testing multiple values on regex101

https://github.com/renovatebot/renovate/blob/5bfa68b8d3bc4cfef94beb5d0d3386e4ed54ca91/lib/workers/global/autodiscover.ts#L33-L52

https://github.com/renovatebot/renovate/blob/5bfa68b8d3bc4cfef94beb5d0d3386e4ed54ca91/lib/util/regex.ts#L54-L91

**Reply by viceice** [reply]
one can use an invert regex `!/.../`

**Reply by PhilipAbed** [reply]
well if we check the regex on the folders one by one then he doesnt need the global flag i guess, it should work without

**Reply by viceice** [reply]
it's probably still needed for regex101 for multiple matches

**Reply by stefandoorn** [reply]
I've played with the regex101 from your example and got what I wanted there. Basically as you did, just other words. All good there.

But whatever I do on the RENOVATE_EXTRA_FLAGS env var, I'm not getting it to work.

> Failed to parse regex pattern "/group\/(?!bodyguard).*/" -> ERROR
> Failed to parse regex pattern "/group\/.*(?!bodyguard)/" -> ERROR

> "stack": "Error: Failed to parse regex pattern \"/group\\/.*(?!bodybuard)/\"\n    at autodiscoverRepositories (/usr/src/app/node_modules/renovate/lib/workers/global/autodiscover.ts:37:15)\n    at processTicksAndRejections (node:internal/process/task_queues:96:5)\n    at Object.start (/usr/src/app/node_modules/renovate/lib/workers/global/index.ts:131:14)\n    at /usr/src/app/node_modules/renovate/lib/renovate.ts:16:22"

Then I played a bit with adding a `/g` or `/i` at the end. Then there is no error, but as well zero matches. 

So e.g. I set it as:

```
--autodiscover=true --autodiscover-filter=/group\/(?!bodyguard)/i
```

With "" around it, same issue. 

```
--autodiscover=true --autodiscover-filter="/group\/(?!bodyguard).*/"
```

No result either, no error as well. 

I tried to look through the regexp util and the autodiscover code, but for me there is nothing obvious there that's causing this.

**Reply by stefandoorn** [reply]
As a validation check, setting it to `group/**` does list all my repositories.

**Reply by viceice** [reply]
your shell is interpreting your regex. you need to properly escape.

better use a config file

**Reply by stefandoorn** [reply]
Can I just move these config items to the `renovate.json` config inside the `renovate-runner` repo I've created following the steps here https://gitlab.com/renovate-bot/renovate-runner? Or is that `renovate.json` specific for updating that repo's dependencies?

**Reply by viceice** [reply]
you can add that to your repo, assuming you don't mirror our repo and include it only in gitlabci CI config yaml.

**Reply by stefandoorn** [reply]
I have a repo with .gitlab-ci.yml:

```
include:
  - project: 'renovate-bot/renovate-runner'
    file: '/templates/renovate-dind.gitlab-ci.yml'

variables:
  RENOVATE_SECRETS: "{\"NPM_READ_TOKEN\": \"${NPM_READ_TOKEN}\"}"

renovate:
  tags:
    - privileged
```

Then I copied the `renovate.json` from here: https://gitlab.com/renovate-bot/renovate-runner/-/blob/main/renovate.json

And added `autoDiscover` and `autoDiscoverFilter` to it. But I feel this is not the way I should be doing it, as nothing get's discovered now.

**Reply by stefandoorn** [reply]
Ah, wait. I might need a `config.js` here.. isn't it?

**Reply by viceice** [reply]
sorry, yes. it can also be `config.json`. the other file is for self update only

**Reply by viceice** [reply]
maybe your need to set the renovate config file environment variable if you don't use the config.js

**Reply by stefandoorn** [reply]
Ah! That part I found a bit confusing in the README here: https://gitlab.com/renovate-bot/renovate-runner.

Added a `config.js` now, just running into the parsing error again:

```
Fatal error: Failed to parse regex pattern "/group/(?!bodybuard).*/"
```

```
module.exports = {
    autodiscover: true,
    autodiscoverFilter: "/group\/(?!bodyguard).*/",
}
```

**Reply by stefandoorn** [reply]
Adjusting to

```
module.exports = {
    autodiscover: true,
    autodiscoverFilter: /group\/(?!bodyguard).*/,
}
```

Another error: 

```
WARN: Config needs migrating
       "originalConfig": {"autodiscover": true, "autodiscoverFilter": {}},
       "migratedConfig": {"autodiscover": true, "autodiscoverFilter": {}}
FATAL: Fatal error: invalid pattern
       "err": {
         "message": "invalid pattern",
         "stack": "TypeError: invalid pattern\n    at assertValidPattern (/usr/src/app/node_modules/renovate/node_modules/minimatch/minimatch.js:1[25](https://gitlab.com/group/renovate/renovate-runner/-/jobs/2693433766#L25):11)\n    at module.exports (/usr/src/app/node_modules/renovate/node_modules/minimatch/minimatch.js:2:3)\n    at /usr/src/app/node_modules/renovate/node_modules/minimatch/minimatch.js:61:19\n    at Array.filter (<anonymous>)\n    at autodiscoverRepositories (/usr/src/app/node_modules/renovate/lib/workers/global/autodiscover.ts:43:31)\n    at processTicksAndRejections (node:internal/process/task_queues:96:5)\n    at Object.start (/usr/src/app/node_modules/renovate/lib/workers/global/index.ts:131:14)\n    at /usr/src/app/node_modules/renovate/lib/renovate.ts:16:22"
```

I feel I'm close haha :-)

**Reply by stefandoorn** [reply]
I actually think it looks really close to this other discussion: https://github.com/renovatebot/renovate/discussions/16472

**Reply by viceice** [reply]
yes, lookahead isn't a valid option

**Reply by stefandoorn** [reply]
Ahhh, yes. Ok, then I can't do what I want here I think. It's no big deal just loading everything, but I thought this would be a nice way excluding certain folders.
